framework/src/systems/NonlinearSystemBase.C

```diff
@@ -3589,6 +3596,56 @@ NonlinearSystemBase::setMooseKSPNormType(MooseEnum kspnorm)
     mooseError("Unknown ksp norm type specified.");
 }
 
+bool
+NonlinearSystemBase::shouldEvaluateFNormBeforeSMO() const
+{
+  // This complies with the legacy behavior: always do the redundant residual evaluation
+  if (_app.parameters().get<bool>("use_legacy_initial_residual_evaluation_bahavior"))
+    return true;
+
+  // If solve_type == LINEAR, there is no need to evaluate the initial residual.
+  if (_fe_problem.solverParams()._type == Moose::ST_LINEAR)
+    return false;
+
+  // If the user explicitly says no to "computing pre-initial residual", then directly return false
+  if (!_use_fnorm0_before_smo)
+    return false;
+
+  // Otherwise, if there are active solution-modifying objects, we compute the pre-initial residual
+  // and use it in the relative convergence check.
+  if (hasActiveSolutionModifyingObjects())
+    return true;
```

Comment:
- correct
- and correct

> (because why would we ... the result should be the same)

That's the hope. But if we are not careful enough in accounting for all the SMOs, the solution may change. For example, what if there is a userobject that modifies the solution vector? And in that case, we will have to rely on the user to set `use_fnorm_before_smo` to the value they want.

Reply:
Okay, so I've modified the logic a little bit (see [87526cd](https://github.com/idaholab/moose/pull/23472/commits/87526cd76974cc6c8df3eec830ecb6a90302d10d)) -- if the user explicitly sets this parameter, we will respect the user's choice regardless of the existence of SMOs.

Reply:
ok that's good because it seems like it would have been ignored before if there weren't the few SMOs we're aware of at the framework level...?

Reply:
That's right.